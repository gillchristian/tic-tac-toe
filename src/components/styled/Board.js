import styled from 'styled-components'

const Board = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  width: ${64*3 + 5*2*3 + 20*2}px;
  margin: 20px;
  padding: 20px;
  border-radius: 14px;
  /* Paper stack box-shadow */
  /* @link: http://jsfiddle.net/chriscoyier/PBzLJ/ */
  background: #eee;
  box-shadow:
      0 1px 1px rgba(0,0,0,0.15),
      0 10px 0 -5px #eee,
      0 10px 1px -4px rgba(0,0,0,0.15),
      0 20px 0 -10px #eee,
      0 20px 1px -9px rgba(0,0,0,0.15);
`

export default Board
