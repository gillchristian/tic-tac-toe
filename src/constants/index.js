export const O = '●'
export const X = '✖'
export const s = ''

export const RESULTS = [
  /1..1..1../,
  /.1..1..1./,
  /..1..1..1/,
  /111....../,
  /...111.../,
  /......111/,
  /1...1...1/,
  /..1.1.1../
]

