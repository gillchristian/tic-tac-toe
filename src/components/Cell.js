import R from 'ramda'

import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { X, O } from '../constants'

const getCellColor = R.compose(
  R.cond([
    [R.equals(X), R.always('palevioletred')],
    [R.equals(O), R.always('tomato')],
    [R.T,         R.always('papayawhip')],
  ]),
  R.prop('mark')
)

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${getCellColor};
  width: 64px;
  height: 64px;
  border-radius: 14px;
  cursor: pointer;
  margin: 5px;
  font-weight: bold;
  font-size: 40px;
  color: white;
  box-shadow: 0 2px 5px gray;
  user-select: none;

  &:hover {
    background-color: ${props => !props.mark && 'lightgray'};
    cursor: ${props => !!props.mark && 'default'};
`

class Cell extends React.Component {
  static propTypes = {
    mark: PropTypes.oneOf([O, X, '']),
    next: PropTypes.oneOf([O, X])
  }

  state = {
    mouseIsOver: false
  }

  mouseEnter = () => {
    this.setState({ mouseIsOver: true })
  }

  mouseLeave = () => {
    this.setState({ mouseIsOver: false })
  }

  render() {
    const {mark, next} = this.props

    return (
      <Wrapper
        mark={mark}
        next={next}
        onClick={this.props.onClick}
        onMouseEnter={this.mouseEnter}
        onMouseLeave={this.mouseLeave}
      >
        {
          mark
            ? mark
            : (this.state.mouseIsOver ? next : '')
        }
      </Wrapper>
    )
  }
}

export default Cell

