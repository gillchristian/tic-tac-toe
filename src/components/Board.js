import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { X, O } from '../constants'
import { getO, getX } from '../reducers/score'

import Cell from './Cell'
import ScoreBoard from './styled/ScoreBoard'
import Score from './styled/Score'


const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;
  width: ${64*3 + 5*2*3 + 20*2}px;
  margin:  ${props => props.isOver ? '0 20px 20px' : '20px' };
  padding: 20px;
  border-radius: 14px;
  /* Paper stack box-shadow */
  /* @link: http://jsfiddle.net/chriscoyier/PBzLJ/ */
  background: #eee;
  box-shadow:
      0 1px 1px rgba(0,0,0,0.15),
      0 10px 0 -5px #eee,
      0 10px 1px -4px rgba(0,0,0,0.15),
      0 20px 0 -10px #eee,
      0 20px 1px -9px rgba(0,0,0,0.15);
`

const Board = ({board, next, score, addMark, isOver}) => (
  <Wrapper isOver={isOver}>
    {
      board.map((mark, key) => (
        <Cell
          key={key}
          mark={mark}
          next={isOver ? '' : next}
          onClick={() => !mark && !isOver && addMark(next, key)}
        />
      ))
    }
    <ScoreBoard>
      <Score>
        <Cell mark={O} next={O} />
        {getO(score)}
      </Score>
      <Score>
        <Cell mark={X} next={X} />
        {getX(score)}
      </Score>
    </ScoreBoard>
  </Wrapper>
)

Board.propTypes = {
  isOver: PropTypes.bool,
  board: PropTypes.arrayOf(PropTypes.string).isRequired,
  next: PropTypes.oneOf([O, X]).isRequired,
  score: PropTypes.shape({
    [X]: PropTypes.number,
    [O]: PropTypes.number
  }).isRequired,
  addMark: PropTypes.func.isRequired
}

export default Board

