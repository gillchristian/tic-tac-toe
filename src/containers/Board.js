import { connect } from 'react-redux'

import Board from '../components/Board'

import { addMark } from '../actions/game'

import { getBoard } from '../reducers/board'
import { isOver } from '../reducers/is-over'
import { getNext } from '../reducers/next'
import { getScore } from '../reducers/score'

const mapStateToProps = (state) => ({
  isOver: isOver(state),
  next: getNext(state),
  board: getBoard(state),
  score: getScore(state)
})

const mapDispatchToProps = {
  addMark
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Board)

