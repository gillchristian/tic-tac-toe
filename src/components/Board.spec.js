import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { X, O, s }  from '../constants'
import Board from './Board'
import Cell from './Cell'

describe('Board', () => {
  it('renders ok', () => {
    const props = {
      addMark: jest.fn(),
      next: X,
      isOver: false,
      board: [
        s, s, X,
        s, O, s,
        s, s, s
      ],
      score: {
        [X]: 0,
        [O]: 0
      }
    };

    const wrapper = shallow(<Board {...props} />);
    const cells = wrapper.find(Cell)

    expect(cells.length).toBe(11) // board: 9, score: 2
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('calls addMark onClick on the board Cells when the cell is empty & game is not over', () => {
    const props = {
      addMark: jest.fn(),
      next: X,
      isOver: false,
      board: [
        s, s, X,
        s, O, s,
        s, s, s
      ],
      score: {
        [X]: 0,
        [O]: 0
      }
    };

    const wrapper = shallow(<Board {...props} />);
    const cells = wrapper.find(Cell)

    cells.first().simulate('click')

    expect(props.addMark).toHaveBeenCalledWith(props.next, 0)
  })

  it('does not call addMark onClick on the board Cells when the cell is has a mark', () => {
    const props = {
      addMark: jest.fn(),
      next: O,
      isOver: false,
      board: [
        X, s, X,
        s, O, s,
        s, s, s
      ],
      score: {
        [X]: 0,
        [O]: 0
      }
    };

    const wrapper = shallow(<Board {...props} />);
    const cells = wrapper.find(Cell)

    cells.first().simulate('click')

    expect(props.addMark).not.toHaveBeenCalledWith()
  })

  it('does not call addMark onClick on the board Cells when the game is over', () => {
    const props = {
      addMark: jest.fn(),
      next: X,
      isOver: true,
      board: [
        s, s, O,
        s, O, X,
        O, s, X
      ],
      score: {
        [X]: 0,
        [O]: 0
      }
    };

    const wrapper = shallow(<Board {...props} />);
    const cells = wrapper.find(Cell)

    cells.first().simulate('click')

    expect(props.addMark).not.toHaveBeenCalledWith()
  })
})

