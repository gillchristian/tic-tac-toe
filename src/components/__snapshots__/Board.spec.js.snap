exports[`Board renders ok 1`] = `
<styled.div
  isOver={false}>
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark="X"
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark="O"
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <Cell
    mark=""
    next="X"
    onClick={[Function]} />
  <styled.div>
    <styled.div>
      <Cell
        mark="O"
        next="O" />
    </styled.div>
    <styled.div>
      <Cell
        mark="X"
        next="X" />
    </styled.div>
  </styled.div>
</styled.div>
`;
